Added escaping to ActiveEnum to allow for non-UAX#31 chars, as well as camel_case related conflicts.

denwong47 opened the pull request:
<!--

Thank you for contributing to this project!

If you need any help please feel free to contact us on Discord: https://discord.com/invite/uCPdDXzbdv
Or, mention our core members by typing `@GitHub_Handle` on any issue / PR

Add some test cases! It help reviewers to understand the behaviour and prevent it to be broken in the future.

-->

## PR Info

<!-- mention the related issue -->
- Closes [Issue #1373](https://github.com/SeaQL/sea-orm/issues/1373)

## Bug Fixes

- Fixes `DeriveActiveEnum` throwing
    - `"" is not a valid identifier` panic if `string_value` consists only of non-UAX#31 compliant characters
    - `the name SomeName is defined multiple times` if two enum values have `string_value` of `someName` and `some_name` respectively
    - `"_0 123" is not a valid identifier` if `string_value` starts with any digit followed by at least one space in the string.

## Details

This adds a function `camel_case_with_escaped_non_xid` to `util.rs` of `sea-orm-macros`, which is a wrapper for `camel_case` to conduct further escaping specifically suitable for Rust Identifiers.

It does so by:
- iterating through each character to check
    - if it is not `_` or whitespace, and
    - if its within `XID_Start` character set for first character, and
    - if its within `XID_Continue` character set for the rest of the string
- if not,
    - it decode the Unicode character to `u32`, then replace original character with its `{:X}` hex representation: e.g. whitespace becomes `0x20`.
- the string is then reconstructed
- if the string is at this point empty,
    - return special string `"__Empty"`.
- otherwise apply `camel_case`.
- the first character is then checked for `numeric`; if so, prepend an `underscore`.

For example,
```rust
#[derive(Clone, Debug, PartialEq, EnumIter, DeriveActiveEnum)]
#[sea_orm(
    rs_type = "String",
    db_type = "String(None)",
    enum_name = "conflicting_string_values"
)]
pub enum ConflictingStringValues {
    #[sea_orm(string_value = "")]
    Member1,
    #[sea_orm(string_value = "$")]
    Member2,
    #[sea_orm(string_value = "$$")]
    Member3,
    #[sea_orm(string_value = "AB")]
    Member4,
    #[sea_orm(string_value = "A_B")]
    Member5,
    #[sea_orm(string_value = "A$B")]
    Member6,
    #[sea_orm(string_value = "0 123")]
    Member7,
}
```
will now produce the following Variant Enum:
```rust
        pub enum ConflictingStringValuesVariant {
            __Empty,
            _0x24,
            _0x240x24,
            Ab,
            A0x5Fb,
            A0x24B,
            _0x300x20123,
        }
```

## See Also
[Valid Rust Identifier talking about XID_Start and XID_Continue](https://doc.rust-lang.org/reference/identifiers.html)

denwong47 commented:
Fixed CI errors. Apologies!

tyt2y3 commented on tests/common/features/sea_orm_active_enums.rs:
```diff
@@ -53,3 +53,36 @@ pub enum MediaType {
+
+#[derive(Debug, Clone, PartialEq, Eq, EnumIter, DeriveActiveEnum, Copy)]
+#[sea_orm(rs_type = "String", db_type = "Enum", enum_name = "pop_os_names_typos")]
+pub enum PopOSTypos {
```
😎

tyt2y3 reviewed (approved):
The implementation looks clean to me, and I especially appreciate the documentation.
Still require Billy's review as he is the primary author of `sea-orm-macros`

tyt2y3 commented:
> For example,
> 
> ```rust
> #[derive(Clone, Debug, PartialEq, EnumIter, DeriveActiveEnum)]
> #[sea_orm(
>     rs_type = "String",
>     db_type = "String(None)",
>     enum_name = "conflicting_string_values"
> )]
> pub enum ConflictingStringValues {
>     #[sea_orm(string_value = "")]
>     Member1,
>     #[sea_orm(string_value = "$")]
>     Member2,
>     #[sea_orm(string_value = "$$")]
>     Member3,
>     #[sea_orm(string_value = "AB")]
>     Member4,
>     #[sea_orm(string_value = "A_B")]
>     Member5,
>     #[sea_orm(string_value = "A$B")]
>     Member6,
>     #[sea_orm(string_value = "0 123")]
>     Member7,
> }
> ```
> 
> will now produce the following Variant Enum:
> 
> ```rust
>         pub enum ConflictingStringValuesVariant {
>             __Empty,
>             _0x24,
>             _0x240x24,
>             Ab,
>             A0x5Fb,
>             A0x24B,
>             _0x300x20123,
>         }
> ```
> 
> ## See Also
> [Valid Rust Identifier talking about XID_Start and XID_Continue](https://doc.rust-lang.org/reference/identifiers.html)

It will be great if we can also include this example as a doc test

denwong47 commented:
Thank you, I'll have a look at doc testing tomorrow!

denwong47 commented:
Changes post-review:
- rebased onto latest `SeaQL:master`
- added unit test for `camel_case_with_escaped_non_uax31` in `sea_orm_macros::util::test` submodule.
    - @tyt2y3 Unfortunately I wasn't able to do these test as a doc test as suggested - since `util` is private to the sub-crate, doc test can't import it as far as I know.
- renamed the main function of `camel_case_with_escaped_non_xid` to `camel_case_with_escaped_non_uax31` - which is a better description of the standard we are trying to check against.
- fixed `Cargo.toml` in `sea-orm-macros` not specifying feature `tests-cfg`, which seems to be required for an existing unit test.

billy1624 commented:
Hey @denwong47, I just fetch from latest master. Now, I'm going to add a few more test cases here and there before merging. Thanks for the efforts!! Great contribution! :)

billy1624 reviewed (approved):
Thanks again for the PR! More PRs are welcomed! See you around on GitHub :P
